src/webgpu/util/copy_to_texture.ts

```diff
@@ -111,250 +42,107 @@ export class CopyToTextureUtils extends GPUTest {
-        );
-        const checkRowF32 = checkF32.subarray(
-          (y * bytesPerRow) / checkF32.BYTES_PER_ELEMENT,
-          bytesPerRow / checkF32.BYTES_PER_ELEMENT
-        );
-        const checkResult = checkElementsBetween(checkRowF32, [
-          i => expRowF32[i] - 0.001,
-          i => expRowF32[i] + 0.001,
-        ]);
-        if (checkResult !== undefined) return `on row ${y}: ${checkResult}`;
+      if (requireColorSpaceConversion) {
+        // WebGPU support 'srgb' as dstColorSpace only.
+        if (srcColorSpace === 'display-p3' && dstColorSpace === 'srgb') {
+          rgba = displayP3ToSrgb(rgba);
+        } else {
+          unreachable();
+        }
       }
-    } else {
-      for (let y = 0; y < height; ++y) {
-        const exp = new Uint8Array(expected.buffer, expected.byteOffset, expected.byteLength);
-        const check = new Uint8Array(actual.buffer, actual.byteOffset, actual.byteLength);
 
-        const checkResult = checkElementsEqual(
-          check.subarray(y * rowPitch, bytesPerRow),
-          exp.subarray(y * bytesPerRow, bytesPerRow)
-        );
-        if (checkResult !== undefined) return `on row ${y}: ${checkResult}`;
+      if (requirePremultiplyAlpha) {
+        rgba.R *= rgba.A;
+        rgba.G *= rgba.A;
+        rgba.B *= rgba.A;
       }
-    }
-    return undefined;
+
+      return rgba;
+    });
   }
 
   doTestAndCheckResult(
     imageCopyExternalImage: GPUImageCopyExternalImage,
     dstTextureCopyView: GPUImageCopyTextureTagged,
-    copySize: GPUExtent3DDict,
-    bytesPerPixel: number,
-    expectedData: Uint8ClampedArray,
-    dstFormat: RegularTextureFormat
+    expTexelView: TexelView,
+    copySize: Required<GPUExtent3DDict>
   ): void {
     this.device.queue.copyExternalImageToTexture(
       imageCopyExternalImage,
       dstTextureCopyView,
       copySize
     );
 
-    const externalImage = imageCopyExternalImage.source;
-    const dstTexture = dstTextureCopyView.texture;
-
-    const bytesPerRow = align(externalImage.width * bytesPerPixel, kBytesPerRowAlignment);
-    const testBuffer = this.device.createBuffer({
-      size: bytesPerRow * externalImage.height,
-      usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST,
-    });
-    this.trackForCleanup(testBuffer);
-
-    const encoder = this.device.createCommandEncoder();
-
-    encoder.copyTextureToBuffer(
-      { texture: dstTexture, mipLevel: 0, origin: { x: 0, y: 0, z: 0 } },
-      { buffer: testBuffer, bytesPerRow },
-      { width: externalImage.width, height: externalImage.height, depthOrArrayLayers: 1 }
-    );
-    this.device.queue.submit([encoder.finish()]);
-
-    this.checkCopyExternalImageResult(
-      testBuffer,
-      expectedData,
-      externalImage.width,
-      externalImage.height,
-      bytesPerPixel,
-      dstFormat
+    const resultPromise = textureContentIsOKByT2B(
+      this,
+      { texture: dstTextureCopyView.texture },
+      copySize,
+      {
+        maxDiffULPs: 0,
```

Comment:
For now, I've put a fixed 0.001 threshold on all of these tests. I found that even the 0.0002 from the original code for float16 wasn't enough, and I didn't dig enough to find out why.

Reply:
I think this might be the color space conversion matrix precision issue. The matrix CPU used is here (https://github.com/gpuweb/cts/blob/main/src/webgpu/util/color_space_conversion.ts#L63). Even the comments said it is copied from ( http://www.brucelindbloom.com/index.html?Eqn_RGB_XYZ_Matrix.html)  but they're different.

Reply:
The old commit here (https://github.com/gpuweb/cts/commit/f9158e4da0a957c86e262761b657db7e8271f6fd#diff-0781362fbbe96f1c4922e18e4ccead52e6c15573ed9bfa091c5b039c940e4d08R270) shows that for fp16, if we handle around 0 values, the other values could be in 1 ULP diff

Reply:
> I think this might be the color space conversion matrix precision issue.

I suspected this as well, especially because the G channel in some test cases is way farther off than the others.

That old commit is exactly why I'm confused I needed a larger threshold than before. Will see what I can figure out.

Reply:
@shaoboyan would you be able to look into the implementation to find out if the relatively-large diff in the G channel is due to a precision issue in the implementation? It's not very important though, so only if you're able to check easily.